crates/turborepo-globwalk/src/lib.rs

```diff
@@ -308,82 +307,60 @@ pub fn globwalk_internal(
-                }
-            } else {
-                glob.walk(&base_path_new)
-                    .not(ex_patterns.clone())
-                    // Per docs, only fails if exclusion list is too large, since we're using
-                    // pre-compiled globs
-                    .unwrap_or_else(|e| {
-                        panic!(
-                            "Failed to compile exclusion globs: {:?}: {}",
-                            ex_patterns, e,
-                        )
-                    })
-                    .filter_map(|entry| {
-                        let span = info_span!(parent: &span, "visit_file", entry = ?entry);
-                        let _enter = span.enter();
-                        match entry {
-                            Ok(entry)
-                                if walk_type == WalkType::Files && entry.file_type().is_dir() =>
-                            {
-                                None
-                            }
-                            Ok(entry) => Some(
-                                AbsoluteSystemPathBuf::try_from(entry.path()).map_err(|e| e.into()),
-                            ),
-                            Err(e) => {
-                                let io_err = std::io::Error::from(e);
-                                match io_err.kind() {
-                                    // Ignore DNE and permission errors
-                                    std::io::ErrorKind::NotFound
-                                    | std::io::ErrorKind::PermissionDenied => None,
-                                    _ => Some(Err(io_err.into())),
-                                }
-                            }
-                        }
-                    })
-                    .collect::<Vec<_>>()
-            }
+        .map(glob_with_contextual_error)
+        .map_ok(|glob| walk_glob(walk_type, &base_path_new, &ex_patterns, glob))
+        // flat map to bring the results in the vec to the same level as the potential outer err
+        // this is the same as a flat_map_ok
+        .flat_map(|s| s.unwrap_or_else(|e| vec![Err(e)]))
+        .collect()
+}
+
+#[tracing::instrument(fields(glob=glob.to_string().as_str()))]
+fn walk_glob(
+    walk_type: WalkType,
+    base_path_new: &PathBuf,
+    ex_patterns: &Vec<Glob>,
+    glob: Glob,
+) -> Vec<Result<AbsoluteSystemPathBuf, WalkError>> {
+    glob.walk(&base_path_new)
+        .not(ex_patterns.clone())
+        .unwrap_or_else(|e| {
+            // Per docs, only fails if exclusion list is too large, since we're using
+            // pre-compiled globs
+            panic!(
+                "Failed to compile exclusion globs: {:?}: {}",
+                ex_patterns, e,
+            )
         })
-        .collect::<Result<HashSet<_>, WalkError>>()?;
-    Ok(result)
+        .filter_map(|entry| visit_file(walk_type, entry))
+        .collect::<Vec<_>>()
+}
+
+#[tracing::instrument]
+fn visit_file(
```

Comment:
If I'm understanding this correctly this is just moved out from `globwalk_internal`?

Reply:
Yeah I decided to pull them out rather than using spans manually because it was a little noisy